subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+
+[[sec:sam_with_receiver]]
+=== `sam-with-receiver` compiler plugin
+
+Gradle comes with the https://docs.gradle.org/current/javadoc/org/gradle/api/HasImplicitReceiver.html[@HasImplicitReceiver] annotation. When used in conjunction with Kotlin https://kotlinlang.org/docs/sam-with-receiver-plugin.html[sam-with-receiver compiler plugin], it can turn the parameter of https://kotlinlang.org/docs/fun-interfaces.html[Single Abstract Method (SAM)] interfaces like https://docs.gradle.org/current/javadoc/org/gradle/api/Action.html[Action] into implicit receiver making your code look more like a DSL:
+
+[source,kotlin]
+----
+// Without sam-with-receiver
+tasks.register("hello") { task ->
+    // You need to use the lambda parameter
+    task.doLast {
+        println("Hello World")
+    }
+}
+
+// With sam-with-receiver
+tasks.register("hello") {
+    // You can use the receiver
+    doLast {
+        println("Hello World")
+    }
+}
+----
+
+The latter looks closer to `build.gradle.kts` scripts and more concise but is more prone to scoping errors as well.
+
+
+[[sec:kotlin_assignment]]
+=== `kotlin-assignment` compiler plugin
+
+The `kotlin-assignment` compiler plugin is still experimental. It allows overriding assignments so that Kotlin callers do not have to call `.set()` on https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Property.html[Property]:
+
+
+----
+// Without kotlin-asignment
+tasks.jar.configure {
+    // You need to call .set()
+    archiveBaseName.set("mylib")
+}
+
+// With kotlin-asignment
+tasks.jar.configure {
+    // No need to call .set()
+    archiveBaseName = "mylib"
+}
+----
+
+[[sec:api_version]]
+=== `apiVersion` compiler option
+
+The https://kotlinlang.org/docs/compiler-reference.html#api-version-version[Kotlin compiler apiVersion option] enforces that the generated bytecode only uses declaration from the given version. This is important to stay compatible with the Kotlin version embedded by Gradle.
+
+Note: while this enforces your code is compatible with a given version of Kotlin stdlib and Gradle, it does not check dependencies, so it is important to check your dependencies are also compatible.
+
+[[sec:sam_conversion_class]]
+=== `-Xsam-conversions=class` compiler option
+
+`-Xsam-conversions=class` sets up the implementation strategy for SAM (single abstract method) conversion to always generate anonymous classes, instead of using the `invokedynamic` JVM instruction. This is to make sure the lambdas are serializable and provide a better support for configuration cache and incremental build. This is particularly important for lambdas used from `doLast:
+See https://github.com/gradle/gradle/issues/17052[Issue #17052] for more information.
+
+[source,kotlin]
+----
+// Without `-Xsam-conversions=class` task will never be up-to-date
+task.doLast {
+    // Something
+}
+----
+
+
+[[sec:plugins_for_your_plugins]]
+== Different development plugins
```

Comment:
The Gradle team needs to do something about this section. With my new user hat on, this isn't very clear. It should be straightforward. Until the Gradle team can improve the situation, this section is still good documentation of the problem.